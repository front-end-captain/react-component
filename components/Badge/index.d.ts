import * as React from "react";

interface BadgeProps {
  prefixCls?: string;
  className?: string;
  style?: React.CSSProperties;
  wrapStyle?: React.CSSProperties;

  // 展示的数字，大于 overCount 时显示为 ${overCount}+，为 0 时隐藏
  // 也可以是自定义图标icon
  count?: React.ReactNode;

  // 展示封顶的数字值
  overCount?: number;

  // 不展示数字，只有一个小红点
  dot?: boolean;

  // 设置状态点的位置偏移，格式为 [right, top]
  offset?: [number, number];

  // 当数值为 0 时，是否展示 Badge
  showZero?: boolean;

  // 设置 Badge 为状态点
  status?: "default" | "success" | "error" | "warn" | "info";

  // 在设置了 status 的前提下有效，设置状态点的文本
  text?: string;

  // 设置鼠标放在状态点上时显示的文字
  title?: string;
}

declare const Badge: (props: BadgeProps) => React.ReactNode;

export default Badge;
