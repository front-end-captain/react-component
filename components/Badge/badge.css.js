import styled from "styled-components";

import {
  defaultBorderColor,
  successColor,
  infoColor,
  errorColor,
  warningColor,
} from "./../_style/vars.js";

const BadgeWrapper = styled.div`
  font-size: 14px;
  font-variant: tabular-nums;
  font-feature-settings: "tnum";
  line-height: 1;
  color: rgba(0, 0, 0, 0.65);
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  list-style: none;
  position: relative;
  display: inline-block;

  &.${(props) => props.prefixCls}-wrapper-status {
    vertical-align: middle;
    line-height: inherit;
  }

  .${(props) => props.prefixCls}-container {
    position: absolute;
    top: ${(props) => props.top}px;
    right: ${(props) => props.right}px;
    height: 20px;
    border-radius: 10px;
    min-width: 20px;
    line-height: 20px;
    text-align: center;
    font-size: 12px;
    font-weight: normal;
    white-space: nowrap;
    z-index: 10;
    transform: translate(50%, -50%);
    transform-origin: 100% 0;
  }

  .${(props) => props.prefixCls}-no-icon {
    background-color: #f5222d;
    color: #fff;
    box-shadow: 0 0 0 1px #fff;
  }

  .${(props) => props.prefixCls}-container-hide {
    display: none;
  }

  .${(props) => props.prefixCls}-container-show {
    display: inline-block;
  }

  .${(props) => props.prefixCls}-no-child {
    position: relative;
    display: block;
    top: auto;
    transform: none;
  }

  .${(props) => props.prefixCls}-multiple-words {
    padding: 0 8px;
  }

  .${(props) => props.prefixCls}-container-dotted {
    height: 6px;
    width: 6px;
    min-width: 6px;
    border-radius: 100%;
    background: #f5222d;
    z-index: 10;
    box-shadow: 0 0 0 1px #fff;
  }

  .${(props) => props.prefixCls}-status-dot {
    width: 6px;
    height: 6px;
    min-width: 6px;
    display: inline-block;
    border-radius: 50%;
    vertical-align: middle;
    position: relative;
  }

  .${(props) => props.prefixCls}-status-text {
    color: rgba(0, 0, 0, 0.65);
    font-size: 14px;
    margin-left: 8px;
  }

  .${(props) => props.prefixCls}-container-default {
    background-color: ${defaultBorderColor};
  }

  .${(props) => props.prefixCls}-container-success {
    background-color: ${successColor};
  }

  .${(props) => props.prefixCls}-container-error {
    background-color: ${errorColor};
  }

  .${(props) => props.prefixCls}-container-warn {
    background-color: ${warningColor};
  }

  .${(props) => props.prefixCls}-container-info {
    background-color: ${infoColor};
  }
`;

export default BadgeWrapper;
