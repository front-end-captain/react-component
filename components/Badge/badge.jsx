import React, { isValidElement } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import BadgeWrapper from "./badge.css.js";

const Badge = (props) => {
  const {
    prefixCls,
    className,
    style,
    count,
    overCount,
    dot,
    offset,
    showZero,
    status,
    text,
    title,
    children,
    wrapStyle,
  } = props;

  const [right, top] = offset;

  const isDot = dot || (status !== null && typeof status === "string");

  const wrapperClassName = classNames(`${prefixCls}-wrapper`, className, {
    [`${prefixCls}-wrapper-status`]: status !== null && typeof status === "string",
  });
  const badgeClassName = classNames(`${prefixCls}-container`, `${prefixCls}-container-${status}`, {
    [`${prefixCls}-container-hide`]: Number(count) === 0 && !showZero && !dot,
    [`${prefixCls}-container-show`]: showZero,
    [`${prefixCls}-container-dotted`]: dot,
    [`${prefixCls}-no-icon`]: !Number.isNaN(Number(count)),
    [`${prefixCls}-no-child`]: children === undefined,
    [`${prefixCls}-multiple-words`]: Number(count).toString().length >= 2,
    [`${prefixCls}-status-dot`]: status !== null && typeof status === "string",
  });

  const getCountNode = (count) => {
    // count 为自定义 icon
    if (Number.isNaN(Number(count))) {
      if (isValidElement(count)) {
        return count;
      }
    }

    return count > overCount ? `${overCount}+` : count;
  };

  return (
    <BadgeWrapper
      className={wrapperClassName}
      style={wrapStyle}
      prefixCls={prefixCls}
      top={top}
      right={right}
    >
      <div className={badgeClassName} style={style} title={title}>
        {!isDot && getCountNode(count)}
      </div>
      {status !== null && typeof status === "string" && text && (
        <span className={`${prefixCls}-status-text`}>{text}</span>
      )}
      {children && <div className={`${prefixCls}-child`}>{children}</div>}
    </BadgeWrapper>
  );
};

const defaultCount = 0;

Badge.defaultProps = {
  prefixCls: "sky-badge",
  count: defaultCount,
  overCount: 99,
  dot: false,
  offset: [0, 0],
  showZero: false,
  status: null,
  text: "",
  title: defaultCount.toString(),
};

Badge.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  style: PropTypes.objectOf(PropTypes.string),
  wrapStyle: PropTypes.objectOf(PropTypes.string),
  count: PropTypes.oneOfType([PropTypes.number, PropTypes.element]),
  overCount: PropTypes.number,
  dot: PropTypes.bool,
  offset: PropTypes.arrayOf(PropTypes.number),
  showZero: PropTypes.bool,
  status: PropTypes.oneOf(["default", "success", "error", "warn", "info"]),
  text: PropTypes.string,
  title: PropTypes.string,
};

export default Badge;
