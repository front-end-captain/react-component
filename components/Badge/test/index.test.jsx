import React from "react";
import { render } from "enzyme";
import toJson from "enzyme-to-json";

import Badge from "./../index.js";

describe("Badge", () => {
  it("should be render some Badge components", () => {
    const wrapper = render(
      <div>
        <Badge />
        <Badge count={88} />
        <Badge dot />
        <Badge status="info" />
        <Badge status="info" text="information" />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render custom icon when set count for a ReactNode", () => {
    const customIcon = <div className="custom-icon">icon</div>;
    const wrapper = render(<Badge count={customIcon} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render correct position when set offset", () => {
    const wrapper = render(<Badge dot offset={[40, 10]} />);

    expect(toJson(wrapper)).toMatchSnapshot();

    // expect(wrapper.find(".sky-badge-container").css("right")).toBe(40);
    // expect(wrapper.find(".sky-badge-container").css("top")).toBe(10);
  });
});
