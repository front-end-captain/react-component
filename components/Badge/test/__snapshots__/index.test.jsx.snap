// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should be render some Badge components 1`] = `
<div>
  <div
    class="sky-badge-wrapper badgecss__BadgeWrapper-cqs0r8-0 bNEfSO"
  >
    <div
      class="sky-badge-container sky-badge-container-null sky-badge-container-hide sky-badge-no-icon sky-badge-no-child"
      title="0"
    >
      0
    </div>
  </div>
  <div
    class="sky-badge-wrapper badgecss__BadgeWrapper-cqs0r8-0 bNEfSO"
  >
    <div
      class="sky-badge-container sky-badge-container-null sky-badge-no-icon sky-badge-no-child sky-badge-multiple-words"
      title="0"
    >
      88
    </div>
  </div>
  <div
    class="sky-badge-wrapper badgecss__BadgeWrapper-cqs0r8-0 bNEfSO"
  >
    <div
      class="sky-badge-container sky-badge-container-null sky-badge-container-dotted sky-badge-no-icon sky-badge-no-child"
      title="0"
    />
  </div>
  <div
    class="sky-badge-wrapper sky-badge-wrapper-status badgecss__BadgeWrapper-cqs0r8-0 bNEfSO"
  >
    <div
      class="sky-badge-container sky-badge-container-info sky-badge-container-hide sky-badge-no-icon sky-badge-no-child sky-badge-status-dot"
      title="0"
    />
  </div>
  <div
    class="sky-badge-wrapper sky-badge-wrapper-status badgecss__BadgeWrapper-cqs0r8-0 bNEfSO"
  >
    <div
      class="sky-badge-container sky-badge-container-info sky-badge-container-hide sky-badge-no-icon sky-badge-no-child sky-badge-status-dot"
      title="0"
    />
    <span
      class="sky-badge-status-text"
    >
      information
    </span>
  </div>
</div>
`;

exports[`Badge should render correct position when set offset 1`] = `
<div
  class="sky-badge-wrapper badgecss__BadgeWrapper-cqs0r8-0 fddydL"
>
  <div
    class="sky-badge-container sky-badge-container-null sky-badge-container-dotted sky-badge-no-icon sky-badge-no-child"
    title="0"
  />
</div>
`;

exports[`Badge should render custom icon when set count for a ReactNode 1`] = `
<div
  class="sky-badge-wrapper badgecss__BadgeWrapper-cqs0r8-0 bNEfSO"
>
  <div
    class="sky-badge-container sky-badge-container-null sky-badge-no-child sky-badge-multiple-words"
    title="0"
  >
    <div
      class="custom-icon"
    >
      icon
    </div>
  </div>
</div>
`;
